Add gamma kernel in topology

babsey opened the pull request:
The purpose of this PR is to implement gamma kernel in topology (See #809).

heplesser commented on topology/topology_parameter.h:
```diff
@@ -43,6 +43,9 @@
 #include "topologymodule.h"
 
+// Includes for gamma kernel:
+#include <gsl/gsl_sf_gamma.h>
```
Move to very top of all includes with comment `// C includes:`

heplesser commented on topology/topology_parameter.h:
```diff
@@ -458,6 +461,52 @@ class Gaussian2DParameter : public TopologyParameter
 
 
 /**
+ * Gamma parameter p(d) = d^(kappa-1)*exp(-d/theta)/(theta^kappa*gamma(kappa))
```
Should this be `Gamma(kappa)`?

heplesser commented on topology/topology_parameter.h:
```diff
@@ -458,6 +461,52 @@ class Gaussian2DParameter : public TopologyParameter
+  raw_value( double x ) const
+  {
+    return std::pow( x, kappa_ - 1 ) * std::exp( -x / theta_ )
+     / ( std::pow( theta_, kappa_ ) * gsl_sf_gamma( kappa_ ) );
```
For efficiency, `1 / ( std::pow( theta_, kappa_ ) * gsl_sf_gamma( kappa_ ) )` should be pre-computed in the constructor so that one only needs a single multiplication here. Also, precompute `beta_ = 1.0 / theta_` in the constructor, so that you avoid the division in the `exp()` argument.

heplesser commented on topology/topology_parameter.h:
```diff
@@ -458,6 +461,52 @@ class Gaussian2DParameter : public TopologyParameter
+  }
+
+private:
+  double kappa_, theta_;
```
Individual declarations please, with doxygen comment of parameters.

heplesser commented:
@babsey Have you had a chance to look at the review comments and the problems occurring on Travis?

babsey commented:
Here, the updated code and test suite for Gamma kernel in topology. Note that I am not quite skilled to write in C++ but I tried to modify according to @heplesser requests.

Parts of equation is pre-computed in the constructor for better efficiency. I also implemented factorial function but in comment in case we want to compile it without GSL. I apologize that I am not aware of Travis testing and failed to find the source of the error. 

For testing suite in pynest, I implemented the python code to compare the connection distributions with Gamma PDF. Here, I do not understand why NEST connect neurons in Gamma shaped manner for kappa + 1 (See the visual comparision).

babsey commented:
Looking at log in Travis I found that this line (3922) probably indicates fails in Travis: 
`../topology/libtopology.so: undefined reference to "gsl_sf_gamma"`

hakonsbm commented:
@babsey Yes, you are right. When NEST is compiled without support for GSL, functions from that library, such as `gsl_sf_gamma()` that's being used in the `GammaParameter`-class, are unavailable. As I pointed out in my review, you need to use an `#ifdef HAVE_GSL` directive, which will tell the compiler to use a block of code only if NEST is compiled with GSL. That way, if there is no GSL support, you can either make the whole class unavailable, or you can implement an alternative way to do the calculation.

Additionally, line 60 and line 97 in `test_topology_gamma_kernel.py` are too long.

hakonsbm commented on pynest/nest/tests/test_topology_gamma_kernel.py:
```diff
@@ -0,0 +1,124 @@
+import unittest
+import nest
+import nest.topology as tp
+from collections import Counter
```
`collections` isn't used in this test.

hakonsbm commented on pynest/nest/tests/test_topology_gamma_kernel.py:
```diff
@@ -0,0 +1,124 @@
+    """Tests of gamma kernel in topology"""
+
+    def test_targets(self):
+        """Object targets"""
```
It should maybe have a more descriptive name?

hakonsbm commented on pynest/nest/tests/test_topology_gamma_kernel.py:
```diff
@@ -0,0 +1,124 @@
+        })
+
+        gamma1 = sts.gamma.pdf(x, a=kappa, scale=theta)
+        gamma_normed1 = gamma1 / float(np.sum(gamma1)) / dx
```
As these are only used when plotting for debugging, they should be moved down to the debug block, or removed.

hakonsbm commented on pynest/nest/tests/test_topology_gamma_kernel.py:
```diff
@@ -0,0 +1,124 @@
+        difference = np.diff([gamma_normed2[:-1], h_normed], axis=0)
+        error_score = np.abs(np.median(difference))
+        # print('Error: %s' % error_score)
+        assert(error_score < 10e-5)
```
I suggest using ```self.assertLess(error_score, 1e-4) ``` here, as it gives better feedback if the assert fails.

hakonsbm commented on pynest/nest/tests/test_topology_gamma_kernel.py:
```diff
@@ -0,0 +1,124 @@
+            ax.step(x, gamma_normed1, label='Gamma PDF: $a = \kappa$')
+            ax.step(x, gamma_normed2, label='Gamma PDF: $a = \kappa + 1$')
+            ax.legend()
+            pl.show()
```
As this block is only used for debugging, I suggest removing `if False` and either commenting the block out, or removing it completely.

hakonsbm commented on topology/topology_parameter.h:
```diff
@@ -456,6 +456,66 @@ class Gaussian2DParameter : public TopologyParameter
+      x *= i;
+    }
+    return x;
+  }
```
This implementation of the factorial only works for integers. However, `kappa_` can be any `double` value. Passing it to the `factorial` function will then implicitly convert it to an integer, and if `kappa_` is a decimal number, we get the wrong value for `delta_`. I suggest instead using the [Spouge's approximation](https://en.wikipedia.org/wiki/Spouge%27s_approximation). You can find an example implementation [here](https://rosettacode.org/wiki/Gamma_function#C).

heplesser commented on pynest/nest/tests/test_topology_gamma_kernel.py:
```diff
@@ -0,0 +1,121 @@
+        h = np.histogram(d, bins=x)[0]
+        h_normed = h / float(np.sum(h)) / dx
+        difference = np.diff([gamma_normed[:-1], h_normed], axis=0)
+        error_score = np.abs(np.median(difference))
```
Instead of this rather *ad hoc* test, why not use `scipy.stats.kstest()` as a proper statistical test?

babsey replied:
That is my concerning. I still do not understand, why PDF of the Gamma distribution function shows not matching results with the distribution of targets as a function of distance calculated in NEST. 

Using the same value in NEST and in Python the distributions scored a low match (Fig. 1-3, blue and orange traces). Then I increased kappa by 1 for PDF in python and compared with the distribution calculated from NEST. Here, these distributions have a better matching score (Fig 1-3, blue and green traces)

In my view, the PDFs in python were calculated correctly but I do not understand why NEST Topology gave different output.

### Figures

![gamma_kappa_1_theta_10](https://user-images.githubusercontent.com/620984/37700500-17ee27a2-2cec-11e8-8bb5-4bae6e1f2bdf.png)
Fig 1.

----------

![gamma_kappa_2_theta_5](https://user-images.githubusercontent.com/620984/37700657-ada047da-2cec-11e8-97cf-ed70674b1dbb.png)
Fig 2.

----------

![gamma_kappa_3_theta_4](https://user-images.githubusercontent.com/620984/37700675-b62d83c2-2cec-11e8-922e-432b7ed2a485.png)
Fig 3.

heplesser replied:
@babsey The number of connections you will find in NEST at a given distance depends not only on the kernel value at that distance, but also on the number of potential targets at that distance; for the test, you thus need to convolve the gamma kernel with this density and then apply the KS-test. For details, see [Daniel Hjertholm, *Statistical tests for connection algorithms for structured neural networks* ](https://brage.bibsys.no/xmlui/handle/11250/189117), esp. Chapter 4.

heplesser commented:
@hakonsbm Could you take another look?
